rpmlint/checks/FileDigestCheck.py

```diff
@@ -9,16 +9,165 @@
 DEFAULT_DIGEST_ALG = 'sha256'
 
 
+# We support different "filters" for calculating file digests for whitelisting
+# purposes. The reasoning behind this is that we want to detect *functional*
+# changes to certain files but we don't want to be bothered by any other
+# changes like changes in comments, copyright headers, whitespace and so on.
+#
+# To avoid such bogus changes showing up in the checker we can filter the
+# target files depending on their file type. Currently we have two common
+# special cases:
+#
+# - shell like files (configuration files, shell scripts, Python
+# scripts, Perl scripts etc) that can contain empty lines or comments
+# introduced by '#'.
+# - XML files for things like D-Bus configuration files or polkit policies.
+# Here also XML style comments and whitespace can occur.
+#
+# The filter classes help in calculating a file digest for a filtered version
+# of the target file that is normalized and therefore stable against the
+# outlined kinds of changes. Since this makes getting the right whitelisting
+# digests hard, a small companion tool in tools/get_whitelisting_digest.py
+# exists.
+
+
+class ShellDigester:
+    """This class performs digest calculation of shell style configuration
+    files or scripts. Certain aspects like comments and whitespace will be
+    filtered out of the digest calculation."""
+
+    def __init__(self, path, halg):
+        self.path = path
+        self.halg = halg
+
+    def _parse_content(self, line_cb):
+        with open(self.path, 'r') as fd:
+            line_nr = 0
+            for line in fd:
+                line_nr += 1
+
+                stripped = line.strip()
+                if not stripped:
+                    # skip empty lines
+                    continue
+                elif line_nr == 1 and stripped.startswith('#!'):
+                    # keep shebang lines instact
+                    pass
+                elif stripped.startswith('#'):
+                    # skip comments
+                    continue
+
+                # don't use the completely stripped version, in Python for
+                # example the indentation is part of the syntax and changes
+                # here could change the meaning
+                line_cb(line.rstrip() + "\n")
+
+    def get_filtered(self):
+        """This just returns the filtered version of the input file."""
+        ret = []
+        self._parse_content(lambda line: ret.append(line))
+
+        return ''.join(ret)
+
+    def get_digest(self):
```

Comment:
Please do this just once as get_filtered should return bytes (or string) and then you just calculate hash of it.

Reply:
If it is about the multiple `get_digest()` methods then switching to generator/`yield` approach might be the better way to avoid having to read in all data at once.